src/decoder/lattice-faster-decoder-combine.cc

```diff
@@ -0,0 +1,1093 @@
+  if (prev_toks_.empty()) {
+    if (!warned_) {
+      KALDI_WARN << "Error, no surviving tokens on frame " << frame;
+      warned_ = true;
+    }
+  }
+
+  BaseFloat adaptive_beam;
+  Token *best_tok = NULL;
+  StateId best_tok_state_id;
+  // "cur_cutoff" is used to constrain the epsilon emittion in current frame.
+  // It will not be updated.
+  BaseFloat cur_cutoff = GetCutoff(prev_toks_, &adaptive_beam,
+                                   &best_tok_state_id, &best_tok);
+  KALDI_VLOG(6) << "Adaptive beam on frame " << NumFramesDecoded() << " is "
+                << adaptive_beam;
+
+
+  // pruning "online" before having seen all tokens
+
+  // "next_cutoff" is used to limit a new token in next frame should be handle
+  // or not. It will be updated along with the further processing.
+  BaseFloat next_cutoff = std::numeric_limits<BaseFloat>::infinity();
+  // "cost_offset" contains the acoustic log-likelihoods on current frame in 
+  // order to keep everything in a nice dynamic range. Reduce roundoff errors.
+  BaseFloat cost_offset = 0.0;
+
+  // First process the best token to get a hopefully
+  // reasonably tight bound on the next cutoff.  The only
+  // products of the next block are "next_cutoff" and "cost_offset".
+  // Notice: As the difference between the combine version and the traditional
+  // version, this "best_tok" is choosen from emittion tokens. Normally, the
+  // best token of one frame comes from an epsilon non-emittion. So the best
+  // token is a looser boundary. We use it to estimate a bound on the next
+  // cutoff and we will update the "next_cutoff" once we have better tokens.
+  // The "next_cutoff" will be updated in further processing.
+  if (best_tok) {
+    cost_offset = - best_tok->tot_cost;
+    for (fst::ArcIterator<FST> aiter(*fst_, best_tok_state_id);
+         !aiter.Done();
+         aiter.Next()) {
+      const Arc &arc = aiter.Value();
+      if (arc.ilabel != 0) {  // propagate..
+        // ac_cost + graph_cost
+        BaseFloat new_weight = arc.weight.Value() + cost_offset -
+            decodable->LogLikelihood(frame, arc.ilabel) + best_tok->tot_cost;
+        if (new_weight + adaptive_beam < next_cutoff)
+          next_cutoff = new_weight + adaptive_beam;
+      }
+    }
+  }
+
+  // Store the offset on the acoustic likelihoods that we're applying.
+  // Could just do cost_offsets_.push_back(cost_offset), but we
+  // do it this way as it's more robust to future code changes.
+  cost_offsets_.resize(frame + 1, 0.0);
+  cost_offsets_[frame] = cost_offset;
+
+  // Build a queue which contains the emittion tokens from previous frame.
+  std::vector<StateId> cur_queue;
+  for (IterType iter = prev_toks_.begin(); iter != prev_toks_.end(); iter++) {
+    cur_queue.push_back(iter->first);
+    iter->second->in_current_queue = true;
+  }
+
+  // Iterator the "cur_queue" to process non-emittion and emittion arcs in fst.
+  while (!cur_queue.empty()) {
+    StateId state = cur_queue.back();
```

Comment:
Processing the queue from the same end you push to is probably not optimal as it would cause you to follow chains of epsilons depth-first, which may require them to be searched again if the costs change later on.  If you use a std::dequeue and use both ends, it would likely be more optimal.